jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/JohnsonShortestPaths.java

```diff
@@ -255,14 +286,28 @@ private void runWithNegativeEdgeWeights(Graph<V, E> g)
         assert g.getType().isDirected();
 
         // create extra graph
-        Graph<V, E> extraGraph = new DirectedPseudograph<>(graph.getEdgeFactory());
+        Graph<V,
+            E> extraGraph = GraphTypeBuilder
+                .<V, E> directed().allowingMultipleEdges(true).allowingSelfLoops(true)
+                .edgeSupplier(graph.getEdgeSupplier()).vertexSupplier(graph.getVertexSupplier())
+                .buildGraph();
 
+        /*
+         * FIXME: After next release, keep only the else clause
+         */
         // add new vertex
-        V s = vertexFactory.createVertex();
-        if (g.containsVertex(s)) {
-            throw new IllegalArgumentException("Invalid vertex factory");
+        V s = null;
+        if (vertexFactory != null) { 
+            s = vertexFactory.createVertex();
+            if (!extraGraph.addVertex(s)) { 
+                throw new IllegalArgumentException("Invalid vertex factory");
+            }
+        } else { 
+            s = extraGraph.addVertex();
+            if (s == null) { 
+                throw new IllegalArgumentException("Invalid vertex supplier");
```

Comment:
I would make this error message more explicit. In most cases it is fine to create a Graph without an explicit vertex supplier, and simply use the addVertex(some object) method. This `JohnsonShortestPaths` implementation suddenly expects a graph with an explicit vertex supplier.
Pity there isn't a way to get passed this requirement.

Reply:
Improved the error message.